Update to 1.4.4

Pop000100 opened the pull request:
- Update to 1.4.4's Localization format
- Undo my memory size changes reading about the IL language behind them. changed to use the int type since it is more performant on standard architectures and not worth the memory savings
- Remove redundant code(Config Labels and tooltips, Journey mode sacrifice sizes of 1, tile drops).
- adjust staff balance accordingly.
- rename and rejigger functions.
- change vanilla tweaks interop(JiT is unhappy with the old way)
- update & document code used vanilla code for autofisher
- do discussed EoL changes for redundant features that are included in 1.4.4.

Pop000100 committed:
Update to 1.4.4

- Update to 1.4.4's Localization format
- Undo my memory size changes reading about the IL language behind them. changed to use the int type since it is more performant on standard architectures and not worth the memory savings
- Remove redundant code(Config Labels and tooltips, Journey mode sacrifice sizes of 1, tile drops).
- adjust staff balance accordingly.
- rename and rejigger functions.
- change vanilla tweaks interop(JiT is unhappy with the old way)
- update & document code used vanilla code for autofisher
- do discussed EoL changes for redundant features that are included in 1.4.4.

gardenappl commented on Items/Consumable/LifeStone.cs:
```diff
@@ -35,16 +30,11 @@ public override void SetDefaults()
 			Item.useAnimation = 30;
 			Item.UseSound = SoundID.Item29;
-			Item.consumable = true;
+			Item.consumable = false;
```
Hmm. I don't remember the difference between `Item.consumable` and `ConsumeItem` at this point. Does this actually work?

Pop000100 replied:
It does with this change

Short Summary:
• Adds item.Consummable as a check in Quick Heal and Quick Mana
• Is a simpler solution then overriding ConsumeItem(


Porting Notes:
• For Quick heal, Quick Mana non-consummables, it is recommended to use Item.consummable field instead of overriding ConsumeItem()
• Existing overrides will continue to work

gardenappl commented on Items/Consumable/ManaStone.cs:
```diff
@@ -11,18 +11,14 @@ namespace GoldensMisc.Items.Consumable
-		byte uses = 0;
-		const byte maxUses = 60;
+		int uses = 0;
+		const int maxUses = 60;
```
What about network packets and save data? Will this be compatible with older save files?

Pop000100 replied:
Yeah I was suppose to cast them which it seems I didn't do

gardenappl commented on GoldensMisc.csproj:
```diff
@@ -10,4 +10,9 @@
   </ItemGroup>
+  <ItemGroup>
+    <Reference Include="VanillaTweaks">
+      <HintPath>..\VanillaTweaks\bin\Debug\net6.0\VanillaTweaks.dll</HintPath>
```
Does this mean that the project now has a hard dependency on VanillaTweaks for the development environment?

Pop000100 replied:
It does see comment below

gardenappl commented on Items/Placeable/SandstoneSlabWall.cs:
```diff
@@ -47,23 +49,24 @@ public override void AddRecipes()
+			}
+		}
+
+        [JITWhenModsEnabled("VanillaTweaks")]
```
This is very interesting, but IMO not worth it if it requires having VanillaTweaks.dll to be able to compile Miscellania. It's not like this code will run very frequently anyway.

Pop000100 replied:
Let me try some more fiddling.

Pop000100 replied:
I got a better solution It is more similar to the old solution. It is contained in the Vanilla tweaks changes. They will be showing up soon(probably early next week)

gardenappl commented on Items/Tools/WormholeHacks.cs:
```diff
@@ -15,18 +15,18 @@ public static class WormholeHacks
 
             //But why the fuck do you not use them sometimes?!
-            On.Terraria.Player.HasItem += PlayerHasItem;
+            Terraria.On_Player.HasItemInInventoryOrOpenVoidBag += PlayerHasItem;
```
Are any other changes needed to support Void Bags? Or do they just work automatically with `player.HasItem`?

Pop000100 replied:
player.HasItem didn't exist anymore, player.HasItemInInventory or that function. I did miss somechanges to HasWormholeItem.

gardenappl replied:
Hmm, I forgot why I hooked into `Player.HasItem` for the wormhole mirror, I should've left a better comment. Last commit looks good for now.

gardenappl commented on Projectiles/AutofisherBobber.cs:
```diff
@@ -16,1724 +16,1286 @@
+                Vector2 Origin = new((float)(Projectile.position.X + Projectile.width * 0.5), (float)(Projectile.position.Y + Projectile.height * 0.5));
+                double OriginXWorldCoord = (te.Position.X + 1.5f) * 16f - Origin.X;
+                double OriginYWorldCoord = (te.Position.Y + 1) * 16f - Origin.Y;
+                float positionSqrt = (float)Math.Sqrt(OriginXWorldCoord * OriginXWorldCoord + OriginYWorldCoord * OriginYWorldCoord);
```
I presume there is Pythagoras going on here and this is some sort of distance?

Pop000100 replied:
If I remember right this is the bobber's offset to position in the water mechanically and out of the water visually.

gardenappl commented on Projectiles/AutofisherBobber.cs:
```diff
@@ -16,1724 +16,1286 @@
+
+                float divdSqr = (float)(productDivisor / (double)positionSqrt);
+                float divdPosX = (float)OriginXWorldCoord * divdSqr;
+                float divdPosY = (float)OriginYWorldCoord * divdSqr;
```
I really appreciate the attempt but to me personally this did not make the code that much clearer.

Pop000100 replied:
Yeah, I didn't really care to figure out what this code is for. I have spent 20+ hours on this file over my time working with it.
Things work, and that is good enough for me. I assume that you spent a hefty chunk as well and I had to update most of it with 1.4.0

gardenappl commented on Projectiles/AutofisherBobber.cs:
```diff
@@ -16,1724 +16,1286 @@
+            autoFisher.rare = Main.rand.NextBool(rareChance);
+            autoFisher.veryrare = Main.rand.NextBool(veryRareChance);
+            autoFisher.legendary = Main.rand.NextBool(legendaryChance);
+            autoFisher.crate = (Main.rand.Next(100) < crateChance);
```
This is great though. I'm glad that this is does not look like decompiled copy-paste anymore.

gardenappl commented on Projectiles/AutofisherBobber.cs:
```diff
@@ -16,1724 +16,1286 @@
+                                    autoFisher.caughtItem = ItemID.Salmon;
+                                    return;
+                                }
+                                autoFisher.caughtItem = ItemID.Bass;
```
All of the IDs... :flushed:

Pop000100 replied:
Took me several hours.

gardenappl commented on Tiles/RedChimneyTE.cs:
```diff
@@ -32,8 +32,8 @@ public override void Update()
 				Vector2 smokeVel = Vector2.Zero;
-				int smokeType = Main.rand.Next(GoreID.ChimneySmoke1, GoreID.ChimneySmoke3);
-				byte rand = (byte)Main.rand.Next(4);
+				int smokeType = Main.rand.Next(GoreID.ChimneySmoke1, GoreID.ChimneySmoke3 + 1);
```
Good catch

Pop000100 replied:
Yeah, forgot Main.Rand is Exclusive

gardenappl commented on Tiles/SandstoneSlabWall.cs:
```diff
@@ -16,10 +15,9 @@ public override bool IsLoadingEnabled (Mod mod)
+			Main.wallLargeFrames[Type] = 1;
 			Main.wallHouse[Type] = true;
 			DustType = DustID.Stone;
-			ItemDrop = ModContent.ItemType<Items.Placeable.SandstoneSlabWall>();
```
How does tModLoader determine tile drops now?

Pop000100 replied:
Here is the porting notes from Tmodloader

`ModTile.RegisterItemDrop to add fallback and manual type+style item drops`
by **jopojelly** has been merged.
**Pull Request:** <https://github.com/tModLoader/tModLoader/pull/3420>
**Arrives in Preview**: `Available now.`
**Arrives in Stable**: `With Stable Release of 1.4.4`
**Breakage**: 🟠 - **Medium**

**Short Summary**:
> - Reworks automatic tile drops introduced in <https://github.com/tModLoader/tModLoader/pull/3210>
> - `ModBlockType.ItemDrop` removed.
> - `TileLoader.GetItemDropFromTypeAndStyle` no longer falls back to tile style 0. A special fallback will be used if the style is not found.
> - `ModTile.RegisterItemDrop` added. Allows manually registering an item drop for a tile style, or as a fallback item. Used to provide drops for styles that don't have a corresponding item that places it.
> - `ModWall.RegisterItemDrop` added. Allows manual registration of wall drop, useful for unsafe walls that drop safe walls.

**Porting Notes**:
> - Run tModPorter. Remove `ItemDrop` lines. If any would be necessary, use `RegisterItemDrop` method to restore the functionality.
> - If mod has already been ported to 1.4.4 and is using `ModTile.GetItemDrops` already, consider using `ModTile.RegisterItemDrop` instead if `ModTile.GetItemDrops` was only used for providing drops for styles without corresponding items. See how `ExampleChest` was changed in the "Files Changed" tab.
> - If mod has already been ported to 1.4.4 and relied on the fallback behavior, use `RegisterItemDrop(itemtypehere)` to continue providing the previous fallback behavior.

seems as if it traverses to the item then back.

gardenappl commented on Items/Equipable/MagnetismRing.cs:
```diff
@@ -31,19 +27,30 @@ public override void SetDefaults()
+					.AddTile(TileID.TinkerersWorkbench)
+					.Register();
+
+				CreateRecipe()
+						.AddIngredient(ItemID.TreasureMagnet)
+						.AddIngredient(ItemID.CelestialMagnet)
+						.AddIngredient(ItemID.GoldRing)
+						.AddTile(TileID.TinkerersWorkbench)
+						.Register();
+			}
```
This recipe shouldn't depend on the Universal Magnet config, probably

Pop000100 replied:
But it would point to a unloaded item?

gardenappl replied:
Well we separated the Universal Magnet config and the Magnetism Ring config. It should be possible to craft the Magnetism Ring without the Universal Magnet by using the vanilla Treasure Magnet instead.

Pop000100 replied:
Yep your right. I wasn't looking hard enough at the name of the file.
I came back from a trip after working on all of this was originally hoping to submit this pr before.

gardenappl reviewed (commented):
Great work, I just have some questions.

Pop000100 committed:
Address Golden's Concerns

- Cast stone variables to save game compatible(I don't really need to do netsend since that is done at runtime).
- Increase stack size to match vanilla.
- Made ancient forges follow vanilla .automatically
- Remove 3rd hook in WormholeHacks since it appears to be redundant since I used it 25ish times in a server without fail.
- I needed to check the Void bag in WormholeHacks.
- Fix ring recipe.
- remove dependency from VanillaTweaks new solution is in said mod.

gardenappl commented:
Looks good to me! Thank you so much!
